File: Sources/kope/graphics5/device.c
#include "device.h"

#ifdef KOPE_DIRECT3D12
#include <kope/direct3d12/device_functions.h>
#endif

#ifdef KOPE_VULKAN
#include <kope/vulkan/device_functions.h>
#endif

#include <assert.h>

void kope_g5_device_create(kope_g5_device *device, const kope_g5_device_wishlist *wishlist) {
	KOPE_G5_CALL2(device_create, device, wishlist);
}

void kope_g5_device_destroy(kope_g5_device *device) {
	KOPE_G5_CALL1(device_destroy, device);
}

void kope_g5_device_set_name(kope_g5_device *device, const char *name) {
	KOPE_G5_CALL2(device_set_name, device, name);
}

void kope_g5_device_create_buffer(kope_g5_device *device, const kope_g5_buffer_parameters *parameters, kope_g5_buffer *buffer) {
#ifdef KOPE_G5_VALIDATION
	buffer->usage_flags = parameters->usage_flags;
#endif
	KOPE_G5_CALL3(device_create_buffer, device, parameters, buffer);
}

void kope_g5_device_create_command_list(kope_g5_device *device, kope_g5_command_list *list) {
	KOPE_G5_CALL2(device_create_command_list, device, list);
}

void kope_g5_device_create_texture(kope_g5_device *device, const kope_g5_texture_parameters *parameters, kope_g5_texture *texture) {
#ifdef KOPE_G5_VALIDATION
	if (kope_g5_texture_format_is_depth(parameters->format)) {
		assert(parameters->dimension != KOPE_G5_TEXTURE_DIMENSION_3D);
	}

	texture->validation_format = parameters->format;
#endif
	KOPE_G5_CALL3(device_create_texture, device, parameters, texture);
}

kope_g5_texture *kope_g5_device_get_framebuffer(kope_g5_device *device) {
	return KOPE_G5_CALL1(device_get_framebuffer, device);
}

void kope_g5_device_create_query_set(kope_g5_device *device, const kope_g5_query_set_parameters *parameters, kope_g5_query_set *query_set) {
	KOPE_G5_CALL3(device_create_query_set, device, parameters, query_set);
}

void kope_g5_device_execute_command_list(kope_g5_device *device, kope_g5_command_list *list) {
	KOPE_G5_CALL2(device_execute_command_list, device, list);
}

void kope_g5_device_create_sampler(kope_g5_device *device, const kope_g5_sampler_parameters *parameters, kope_g5_sampler *sampler) {
	KOPE_G5_CALL3(device_create_sampler, device, parameters, sampler);
}

void kope_g5_device_create_raytracing_volume(kope_g5_device *device, kope_g5_buffer *vertex_buffer, uint64_t vertex_count, kope_g5_buffer *index_buffer,
                                             uint32_t index_count, kope_g5_raytracing_volume *volume) {
	KOPE_G5_CALL6(device_create_raytracing_volume, device, vertex_buffer, vertex_count, index_buffer, index_count, volume);
}

void kope_g5_device_create_raytracing_hierarchy(kope_g5_device *device, kope_g5_raytracing_volume **volumes, kinc_matrix4x4_t *volume_transforms,
                                                uint32_t volumes_count, kope_g5_raytracing_hierarchy *hierarchy) {
	KOPE_G5_CALL5(device_create_raytracing_hierarchy, device, volumes, volume_transforms, volumes_count, hierarchy);
}

void kope_g5_device_wait_until_idle(kope_g5_device *device) {
	KOPE_G5_CALL1(device_wait_until_idle, device);
}

uint32_t kope_g5_device_align_texture_row_bytes(kope_g5_device *device, uint32_t row_bytes) {
	return KOPE_G5_CALL2(device_align_texture_row_bytes, device, row_bytes);
}

File: Backends/Graphics5/Metal/Sources/kope/metal/device.m
#include "device_functions.h"

#include "metalunit.h"

#include <kope/graphics5/device.h>
#include <kope/util/align.h>

#include <kinc/log.h>
#include <kinc/window.h>

#include <assert.h>

void kope_metal_device_create(kope_g5_device *device, const kope_g5_device_wishlist *wishlist) {
	id<MTLDevice> metal_device = MTLCreateSystemDefaultDevice();
	getMetalLayer().device = metal_device;
	device->metal.device = (__bridge_retained void *)metal_device;
	device->metal.library = (__bridge_retained void *)[metal_device newDefaultLibrary];
}

void kope_metal_device_destroy(kope_g5_device *device) {}

void kope_metal_device_set_name(kope_g5_device *device, const char *name) {}

void kope_metal_device_create_buffer(kope_g5_device *device, const kope_g5_buffer_parameters *parameters, kope_g5_buffer *buffer) {
	id<MTLDevice> metal_device = (__bridge id<MTLDevice>)device->metal.device;
	MTLResourceOptions options = MTLResourceCPUCacheModeWriteCombined;
#ifdef KINC_APPLE_SOC
	options |= MTLResourceStorageModeShared;
#else
	if (gpuMemory) {
		options |= MTLResourceStorageModeManaged;
	}
	else {
		options |= MTLResourceStorageModeShared;
	}
#endif
	id<MTLBuffer> metal_buffer = [metal_device newBufferWithLength:parameters->size options:options];
	buffer->metal.buffer = (__bridge_retained void *)metal_buffer;
}

void kope_metal_device_create_command_list(kope_g5_device *device, kope_g5_command_list_type type, kope_g5_command_list *list) {
	id<MTLDevice> metal_device = (__bridge id<MTLDevice>)device->metal.device;
	id<MTLCommandQueue> command_queue = [metal_device newCommandQueue];
	list->metal.command_queue = (__bridge_retained void *)command_queue;
	list->metal.command_buffer = (__bridge_retained void *)[command_queue commandBuffer];
}

void kope_metal_device_create_texture(kope_g5_device *device, const kope_g5_texture_parameters *parameters, kope_g5_texture *texture) {}

static kope_g5_texture framebuffer;

kope_g5_texture *kope_metal_device_get_framebuffer(kope_g5_device *device) {
	CAMetalLayer *metal_layer = getMetalLayer();
	id<CAMetalDrawable> drawable = [metal_layer nextDrawable];
	framebuffer.metal.texture = (__bridge_retained void *)drawable.texture;
	return &framebuffer;
}

kope_g5_texture_format kope_metal_device_framebuffer_format(kope_g5_device *device) {
	return KOPE_G5_TEXTURE_FORMAT_BGRA8_UNORM;
}

void kope_metal_device_execute_command_list(kope_g5_device *device, kope_g5_command_list *list) {
	id<MTLCommandBuffer> command_buffer = (__bridge id<MTLCommandBuffer>)list->metal.command_buffer;
	[command_buffer commit];

	id<MTLCommandQueue> command_queue = (__bridge id<MTLCommandQueue>)list->metal.command_queue;
	command_buffer = [command_queue commandBuffer];
	list->metal.command_buffer = (__bridge_retained void *)[command_queue commandBuffer];
}

void kope_metal_device_wait_until_idle(kope_g5_device *device) {}

void kope_metal_device_create_descriptor_set(kope_g5_device *device, uint32_t descriptor_count, uint32_t dynamic_descriptor_count,
                                             uint32_t bindless_descriptor_count, uint32_t sampler_count, kope_metal_descriptor_set *set) {}

void kope_metal_device_create_sampler(kope_g5_device *device, const kope_g5_sampler_parameters *parameters, kope_g5_sampler *sampler) {}

void kope_metal_device_create_raytracing_volume(kope_g5_device *device, kope_g5_buffer *vertex_buffer, uint64_t vertex_count, kope_g5_buffer *index_buffer,
                                                uint32_t index_count, kope_g5_raytracing_volume *volume) {}

void kope_metal_device_create_raytracing_hierarchy(kope_g5_device *device, kope_g5_raytracing_volume **volumes, kinc_matrix4x4_t *volume_transforms,
                                                   uint32_t volumes_count, kope_g5_raytracing_hierarchy *hierarchy) {}

void kope_metal_device_create_query_set(kope_g5_device *device, const kope_g5_query_set_parameters *parameters, kope_g5_query_set *query_set) {}

uint32_t kope_metal_device_align_texture_row_bytes(kope_g5_device *device, uint32_t row_bytes) {
	return 0;
}

void kope_metal_device_create_fence(kope_g5_device *device, kope_g5_fence *fence) {}

void kope_metal_device_signal(kope_g5_device *device, kope_g5_command_list_type list_type, kope_g5_fence *fence, uint64_t value) {}

void kope_metal_device_wait(kope_g5_device *device, kope_g5_command_list_type list_type, kope_g5_fence *fence, uint64_t value) {}
